### ISSUE TITLE
Fails to read attribute, but succeeds if loaded in second place

### ISSUE THREAD

**Issue by ferdonline** [BlueBrain/libsonata#139]
I found a very unexpected behavior with the library. It fails if one does:

```python
import libsonata
filename = "/gpfs/bbp.cscs.ch/project/proj12/jenkins/cellular/circuit-1k/nodes.h5"
pop = libsonata.NodeStorage(filename).open_population("All")
pop.get_attribute('region', libsonata.Selection((1,2)) )
```

However it succeeds if one gets another attribute before 'region', like morphology, i.e., after restarting the interpreter:
```c++
(...)
pop.get_attribute('morphology', libsonata.Selection((1,2)) )
pop.get_attribute('region', libsonata.Selection((1,2)) )
```

The only difference of region w.r.t. other attributes seems to be the type, I32 instead if U64. Maybe there's some pop count which is initialized and cached, and doesn't work properly with int32?

**Comment by mgeplf** [comment]
Which version are you using? Is it deployed in spack? If so, what module?

**Comment by mgeplf** [comment]
Looking at the `nodes.h5` file, as you noted `region`'s type is `i32` instead of an unsigned type (like etype, morphology, mtype, synapse_class).  Since they are indices into a the respective `@library` datasets (ie: @library/region), I don't think it makes 
sense to have signed types.

My guess is that HighFive/hdf5 doesn't do the conversion for `i32` -> `size_t` required at this line:
https://github.com/BlueBrain/libsonata/blob/master/src/population.cpp#L267

What created this file?

**Comment by sergiorg-hpc** [comment]
@mgeplf The version we used is 0.1.8, the one deployed on BB5.

Just curious. Even if the file is corrupt somehow, what I wonder is, how could it work if you ask first for something else before asking for the region attribute? The issue can be reproduced with NodeSets as well (i.e., you cannot filter by region, unless you asked for something else first).

**Comment by mgeplf** [comment]
> how could it work if you ask first for something else before asking for the region? 

IIRC, hdf5 does weird conversion stuff, and depending on whether a particular loader has been 'activated' at some point, then it works.  It's strange.

**Comment by sergiorg-hpc** [comment]
Just a short update. I've spent the whole day, a good part of it with Nico, and there is something off with how HDF5 is utilized internally.

I might have found something that leads us to the problem, but I need Nico's help to understand if it is actually the reason or it is just a coincidence.

**Comment by ferdonline** [comment]
Looking up the issue with @sergiorg-hpc we noticed that indeed it can be traced back to hdf5 being unable to load the converter among different charsets. 
Nevertheless we plan to fix the issue by telling HighFive that memory has the same charset than datasets, avoiding conversions.

**Comment by mgeplf** [comment]
To clarify, what tool created this file?

**Comment by sergiorg-hpc** [comment]
@matz-e believes it was created with the `mvdtool` conversion script most likely.

**Comment by sergiorg-hpc** [comment]
See https://github.com/BlueBrain/HighFive/pull/428 for more information.

**Comment by mgeplf** [comment]
Just going to reopen it, so we know to upgrade once the fixes are in.  Thanks for tracking this down, and the updates.

**Comment by sergiorg-hpc** [comment]
This has been fixed in HighFive already, but the package has not been updated yet.

More info about the HighFive PR: https://github.com/BlueBrain/HighFive/pull/428

**Comment by mgeplf** [comment]
Cool, when things are released, just make a PR here to bump the git submodule.  Thanks.

**Comment by mgeplf** [comment]
Has this been fixed?  Should there be an update?

**Comment by sergiorg-hpc** [comment]
We had today the bi-weekly dev-meeting of HighFive and discussed a bit this topic. In order for this change to reach deployment, we need `libsonata` to use the latest HighFive version available.

There is some on-going work with summation reports et al., and after that is merged, we can check again to confirm that the fix is finally available.

I'll keep this opened and get back to you as soon as I know more.

**Comment by mgeplf** [comment]
@sergiorg-hpc any updates on this?

**Comment by sergiorg-hpc** [comment]
@mgeplf Let me bring this up in the next HighFive dev-meeting, ok? I have been busy with random things for the past month and didn't manage to follow-up on this, my apologies.

**Comment by mgeplf** [comment]
Should be fixed in #156; closing this.
